### DISCUSSION TITLE
Dynamic number of composed resources / Terraform "for loop" parity?

### DISCUSSION THREAD

**Discussion by jaylevin** [crossplane/crossplane#3777] [Q&A] [answered]
Is there an existing feature in Crossplane Compositions that allows for a dynamic number of composed resources, similar to how Terraform's "for loops" work -- allowing the user to pass a list into the module, resulting in a new resource to be created for each element in the list.

**Use case:**
I want users (developers) to be able to specify a list of emails in their XRC.

For each email in the list, I want a separate managed resource to be created as part of the XRC.

For example, in the Composition/Claim spec that the developer specifies:
```yaml
users:
  - user1@gmail.com
  - user2@gmail.com
```

Resulting in 2 managed `User` resources to be created and reconciled (adding the users via the provider API)

I can see how I can do this with Helm, but reading the Crossplane Composition documentation I was unable to find anything regarding this use-case.

Is it possible?

**Comment by bobh66** [comment, marked as answer]
The Composition Functions alpha feature in 1.11 is intended to provide this type of functionality

**Reply by G-Salviati** [reply]
Hello, is there any function providing this feature? I would like to patch and transform a dynamic number of resource specification (e.g: a list of gcp projects spec)

**Reply by bobh66** [reply]
There are no plans to add this functionality to patch and transform.  Other solutions like function-go-templating and function-kcl support dynamic resource creation.

**Reply by G-Salviati** [reply]
Oh ok thank you! I these days I've read the concepts section of the docs and I tought that using patch and transform was the preferrable method to write compositions. What's the best practice?

**Reply by bobh66** [reply]
Go templating and KCL are popular, and some people write entirely custom functions in go or python.  I would choose the option that you are most comfortable with.
